src/main/java/com/northconcepts/datapipeline/foundations/examples/pipeline/CustomPipelineAction.java

```diff
@@ -32,56 +40,75 @@ public static void main(String[] args) {
             .addAction(new UpcaseFieldsAction().addField("Country"));
         
         pipeline.run();
+        
+        System.out.println("Generated Code:" + System.lineSeparator());
+        System.out.println(pipeline.getJavaCode().getSource());
     }
     
-}
+    static class UpcaseFieldsAction extends PipelineAction {
 
-class UpcaseFieldsAction extends PipelineAction {
+        private final List<String> fieldNameList;
 
-    private final List<String> fieldNameList;
+        protected UpcaseFieldsAction() {
+            super("transform-upcase-fields", "Upper Case Fields");
+            fieldNameList = new ArrayList<>();
+        }
 
-    protected UpcaseFieldsAction() {
-        super("transform-upcase-fields", "Upper Case Fields");
-        fieldNameList = new ArrayList<>();
-    }
+        @Override
+        public DataReader apply(DataReader reader) throws Throwable {
+            TransformingReader transformingReader = new TransformingReader(reader)
+                    .add(new BasicFieldTransformer(fieldNameList.toArray(new String[fieldNameList.size()])).upperCase());
+            return transformingReader;
+        }
 
-    @Override
-    public DataReader apply(DataReader reader) throws Throwable {
-        TransformingReader transformingReader = new TransformingReader(reader)
-                .add(new UpcaseField(fieldNameList));
-        return transformingReader;
-    }
+        public UpcaseFieldsAction addField(String fieldName) {
+            fieldNameList.add(fieldName);
+            return this;
+        }
 
-    public UpcaseFieldsAction addField(String fieldName) {
-        fieldNameList.add(fieldName);
-        return this;
-    }
+        @Override
+        public void fromRecord(Record source) {
+            if(source.containsField("fieldNameList")) {
+                ArrayValue array = source.getField("fieldNameList").getValueAsArray();
+                for (int i = 0; i < array.size(); i++) {
+                    addField(array.getValueAsRecord(i).getValueAsString());
+                }
+            }
+        }
 
-    @Override
-    public void fromRecord(Record source) {
-    }
+        @Override
+        public Record toRecord() {
+            Record record = new Record();
+
+            if (Util.isNotEmpty(fieldNameList)) {
```

Comment:
Let's reduce the code by emitting an empty array if no fields are in `fieldNameList`.

Reply:
@DeleTaylor you meant I should remove this condition, right?

Reply:
I have removed this condition. If no field, then empty array will be added in record.